File: 0115-distinct-subsequences/0115-distinct-subsequences.py

class Solution:
    def __init__(self):
        self.dp = [[-1] * 1001 for _ in range(1001)]

    def solve(self, s, t, m, n):
        if n == 0:
            self.dp[m][n] = 1
            return 1

        if m == 0:
            self.dp[m][n] = 0
            return 0

        if self.dp[m][n] != -1:
            return self.dp[m][n]

        if s[m - 1] == t[n - 1]:
            self.dp[m][n] = (
                self.solve(s, t, m - 1, n - 1) +
                self.solve(s, t, m - 1, n)
            )
        else:
            self.dp[m][n] = self.solve(s, t, m - 1, n)

        return self.dp[m][n]

    def numDistinct(self, s, t):
        m = len(s)
        n = len(t)

        self.dp = [[-1] * (n + 1) for _ in range(m + 1)]

        return self.solve(s, t, m, n)

File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    typedef unsigned long long ull;
    int numDistinct(string s, string t) {
        int m = s.length();
        int n = t.length();

        // dp array 
        vector<vector<ull>> dp(m+1,vector<ull>(n+1));

        // base case 1 if n==0 return 1
        for(int i =0;i<=m;i++){
            dp[i][0]=1;
        }
        //base case 2 if m ==0 return 0
        for(int i =1 ;i<=n;i++){
            dp[0][i]=0;
        }

        for(int i =1;i<=m;i++){
            for(int j =1;j<=n;j++){
                if(s[i-1]==t[j-1]){
                    dp[i][j]=dp[i-1][j-1] + dp[i-1][j];
                }
                else{
                    dp[i][j]= dp[i-1][j];
                }
            }
        }
        return dp[m][n];
    }
};